sound/soc/sof/trace.c

```diff
@@ -13,11 +13,180 @@
+
+static int trace_filter_append_elem(int32_t key, int32_t value,
+				    struct sof_ipc_trace_filter_elem **elem_list,
+				    size_t *capacity, size_t *counter)
+{
+	struct sof_ipc_trace_filter_elem *elem_ptr;
+	size_t new_capacity;
+
+	/* assert enough space, allocate space for a few more elements to reduce realloc() calls */
+	if (*counter >= *capacity) {
+		new_capacity = *capacity + TRACE_FILTER_ELEMENT_EXPANSION;
+		elem_ptr = krealloc(*elem_list, new_capacity * sizeof(*elem_ptr), GFP_KERNEL);
+		if (!elem_ptr)
+			return -ENOMEM;
+		*elem_list = elem_ptr;
+		*capacity = new_capacity;
+	}
+
+	elem_ptr = *elem_list + *counter;
+	elem_ptr->key = key;
+	elem_ptr->value = value;
+	*counter += 1;
+
+	return 0;
+}
+
+static int trace_filter_parse_entry(struct snd_sof_dev *sdev, const char *line,
+				    struct sof_ipc_trace_filter_elem **elem,
+				    size_t *capacity, size_t *counter)
+{
+	int len = strlen(line);
+	size_t cnt = *counter;
+	int log_level;
+	int uuid_id;
+	int pipe_id;
+	int comp_id;
+	int read;
+	int ret;
+
+	ret = sscanf(line, " %d %x %d %d %n", &log_level, &uuid_id, &pipe_id, &comp_id, &read);
+	if (ret != 4 || read != len) {
+		dev_err(sdev->dev, "error: invalid trace filter entry '%s'\n",
+			line);
+		return -EINVAL;
+	}
+
+	if (uuid_id > 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_UUID,
+					       uuid_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+	if (pipe_id >= 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_PIPE,
+					       pipe_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+	if (comp_id >= 0) {
+		ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_PIPE,
+					       comp_id, elem, capacity, &cnt);
+		if (ret)
+			return ret;
+	}
+
+	ret = trace_filter_append_elem(SOF_IPC_TRACE_FILTER_ELEM_LEVEL |
+				       SOF_IPC_TRACE_FILTER_ELEM_FIN,
+				       log_level, elem, capacity, &cnt);
+	if (ret)
+		return ret;
+
+	/* update counter only when parsing whole entry passed */
+	*counter = cnt;
+
+	return len;
+}
+
+static int trace_filter_parse(struct snd_sof_dev *sdev, char* string,
+			      size_t len, size_t *out_elem_cnt,
+			      struct sof_ipc_trace_filter_elem **out)
+{
+	struct sof_ipc_trace_filter_elem *elems;
+	char *entry = string;
+	char *entry_end;
+	size_t capacity;
+	size_t cnt = 0;
+	int entry_len;
+	int ret = 0;
+
+	if (!string)
+		return -EINVAL;
+
+	/*
+	 * estimate capacity, default entry length
+	 * ('log_level UUID_id pipe_id comp_id' - e.g. '0 1FFFA000 0 0;') is 15,
+	 * each entry consist up to 4 elements, so each element it's about 4 chars.
+	 */
+	capacity = len / TRACE_FILTER_CHARS_PER_ELEMENT_ESTIMATION + 1;
+	elems = kmalloc(capacity * sizeof(*elems), GFP_KERNEL);
+	if (!elems)
+		return -ENOMEM;
+
+	/* split input string by ';', and parse each entry separately in trace_filter_parse_entry */
+	while (entry < string + len) {
+		entry_end = strchrnul(entry, ';');
+		*entry_end = '\0';
```

Comment:
strchrnul returns pointer to entry null terminator when ';' can't be found

Reply:
I know, but are you sure there's a `'\0'` terminator there? You're reading a buffer from the user-space. The user could cat a file to the debugfs node, with no terminating null character and no `';'`

Reply:
so I assert null termination during string copy from user space

Reply:
sorry, could you point out where it is done?

Reply:
```
static ssize_t sof_dfsentry_trace_filter_write(struct file *file,
		const char __user *from, size_t count, loff_t *ppos)
{
        // ...
	/* assert null termination */
	string = kzalloc(count + 1, GFP_KERNEL);

Reply:
@ktrzcinx oh, right, sorry. It seems a bit of an overkill to me to zero the whole buffer, of which all but one byte will be overwritten, instead of just setting that one byte, but this works too of course.

Reply:
changed to kmalloc and explicit null termination